src/Compilers/CSharp/Portable/Symbols/TypeSymbolWithAnnotations.cs

```diff
@@ -20,6 +20,41 @@ internal enum NullableAnnotation : byte
         Nullable,     // Explicitly set by flow analysis
     }
 
+    /// <summary>
+    /// The nullable state of an rvalue computed in <see cref="NullableWalker"/>.
+    /// When in doubt we conservatively use <see cref="NullableFlowState.NotNull"/>
+    /// to minimize diagnostics.
+    /// </summary>
+    internal enum NullableFlowState : byte
+    {
+        NotNull,
+        MaybeNull
+    }
+
+    /// <summary>
+    /// A type and its corresponding flow state resulting from evaluating an rvalue expression.
+    /// </summary>
+    [DebuggerDisplay("{GetDebuggerDisplay(), nq}")]
+    internal readonly struct TypeWithState
+    {
+        public TypeSymbol Type { get; }
+        public NullableFlowState State { get; }
+        public bool HasNullType => Type is null;
+        public bool MaybeNull => State == NullableFlowState.MaybeNull;
+        public bool NotNull => State == NullableFlowState.NotNull;
+        public static TypeWithState ForType(TypeSymbol type) => new TypeWithState(type, type?.CanContainNull() == true ? NullableFlowState.MaybeNull : NullableFlowState.NotNull);
+        public TypeWithState(TypeSymbol type, NullableFlowState state) => (Type, State) = (type, state);
+        public void Deconstruct(out TypeSymbol type, out NullableFlowState state) => (type, state) = (Type, State);
+        public string GetDebuggerDisplay() => $"{{Type:{Type?.GetDebuggerDisplay()}, State:{State}{"}"}";
+        public TypeWithState WithNonNullState() => new TypeWithState(Type, NullableFlowState.NotNull);
```

Comment:
Consider: `WithNotNullState`. The rest of the PR seems to use `NotNull` vs. `NonNull`

Reply:
Also the public API will use `NotNull`.

---
In reply to: [260401044](https://github.com/dotnet/roslyn/pull/33648#discussion_r260401044) [](ancestors = 260401044)

Reply:
I should let @jcouv and @jaredpar fight it out. Changing it back to WithNotNullState.

---
In reply to: [260895414](https://github.com/dotnet/roslyn/pull/33648#discussion_r260895414) [](ancestors = 260895414,260401044)